version: "2.21"

services:
  redis:
    image: redis:latest
    ports:
      - 6379:6379
    restart: on-failure

  db:
    image: postgres:15.0
    environment:
      POSTGRES_DB: ${POSTGRES_DB}
      POSTGRES_USER: ${POSTGRES_USER}
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
    volumes:
      - ./data/grocery.sql:/grocery.sql
      - ./restore.sh:/docker-entrypoint-initdb.d/restore.sh
      - db-data:/var/lib/postgresql/data
    env_file:
      - .env
    restart: on-failure

  web:
    build: ./
    command: gunicorn grocery_sales_forecasting_web.wsgi:application --bind 0.0.0.0:8000 --timeout 90
    ports:
      - 8000:8000
    depends_on:
      db:
        condition: service_started
    env_file:
      - .env
    restart: on-failure
    image: sales-forecasting:cp3

  nginx:
    build: ./nginx
    ports:
      - 1337:80
    depends_on:
      web:
        condition: service_started
    restart: on-failure

  worker:
    command: celery -A grocery_sales_forecasting_web worker --loglevel=INFO -Q train,infer
    depends_on:
      db:
        condition: service_started
      web:
        condition: service_started
      redis:
        condition: service_started
    env_file:
      - .env
    restart: on-failure
    image: sales-forecasting:cp3

volumes:
  db-data: